Issue 4: Added a button allowing to fix different errors and warnings automatically

SebastianZ opened the pull request:
I've added a button _Fix matches_, which allows to fix several obvious errors.

It doesn't fix all errors, though, to avoid false positives. E.g. the `style` attribute is not automatically removed yet, because there are cases like in [`box-shadow`](https://developer.mozilla.org/en-US/docs/Web/CSS/box-shadow) where there are intentionally styles assigned to some elements.
Also, some don't have automatic fixing yet but should have.

Currently there is no indication about which tests provide auto-fixing and which ones don't. If we decide to expose this info to users (maybe with single-error-fixing), we should file a follow-up issue for it.

This PR doesn't include unit tests for the auto-fixing functions. If you want me, I'll try to create them for it, otherwise I'll file a separate issue for them.

Sebastian

SebastianZ commented:
Rebased. @Elchi3 can you please do a review?

Sebastian

Elchi3 commented:
Installed a version with this branch for functional testing this week. Will look at the code at some point as well.

Elchi3 commented:
Found a first data loss with the code-in-pre auto fix:

`<pre class="syntaxbox"><code><var>arr</var>.keys()</code></pre>` is corrected to 
`<pre class="syntaxbox">.keys()</pre>`

SebastianZ commented:
> Found a first data loss with the code-in-pre auto fix:
> 
> `<pre class="syntaxbox"><code><var>arr</var>.keys()</code></pre>` is corrected to 
> `<pre class="syntaxbox">.keys()</pre>`

Thank you for the review!
Fixed in https://github.com/Elchi3/mdn-doc-tests/pull/160/commits/4f05412775b4959a6ee7fe4a6a917116056a15af.

Sebastian

Elchi3 commented:
I wonder if the "Fix matches" button should also re-run the tests, as I click the "Run tests" button immediately after to see the difference right now. Otherwise there is no visual hint that the "fix matches" button did anything.

SebastianZ commented:
> I wonder if the "Fix matches" button should also re-run the tests, as I click the "Run tests" button immediately after to see the difference right now. Otherwise there is no visual hint that the "fix matches" button did anything.

The fixed test should already be re-run as of https://github.com/Elchi3/mdn-doc-tests/pull/160/commits/93a805678554bdbac52bfc15f43732c8f868a1e1#diff-2833cd50a1a5514fd0070224be4bc490R18, though it looks like it's not working. I'll investigate.

Sebastian

SebastianZ commented:
I think I've fixed the error with https://github.com/Elchi3/mdn-doc-tests/pull/160/commits/da8adf2f19635de512dbaade2df324bb188b5861.

Sebastian

Elchi3 commented:
Sorry that this is taking so long. The pull request queue is now empty, so once this is rebased I will try to review this before any new PR.

SebastianZ commented:
The rebase is done and I added auto-fixing for incorrectly wrapped sidebar macros.

What I'd like to add before we merge the PR is auto-fixing functionality for data macro notes and especially some unit tests covering the auto-fixing features.

Sebastian

Elchi3 commented on data/sidebar.html:
```diff
@@ -15,5 +15,6 @@
     </div>
     <progress id="testProgress" max="100"></progress>
     <ul id="tests"></ul>
+    <button id="fixMatches" data-l10n-id="fix_matches">Fix matches</button>
```
Nit: "Fix issues" sounds better to me.

SebastianZ commented:
> > Nit: "Fix issues" sounds better to me.

Changed that. Thanks for the hint!

> I've worked with this for the last few days and had no major issue. I've tested each of the auto-fix tests manually and they did the right thing.

Good to hear!

> I filed one general bug for API headlines, but the rest seems to be working just fine. Great work!

Thank you!

I've also added unit tests for the new functionality now.
With the unit testing I found two issues. I've already fixed one in I've found one issue, which I fixed in https://github.com/Elchi3/mdn-doc-tests/pull/160/commits/36e8dba1e21354924703fdae1698df5c5265fd26. Another one is related to line highlighting in code, for which I'll file an issue.

The rest should work fine. So I guess it's ok to finally merge the changes.

Sebastian

SebastianZ commented:
> Another one is related to line highlighting in code, for which I'll file an issue.

That was actually part of this issue. I've fixed it in https://github.com/Elchi3/mdn-doc-tests/pull/160/commits/cb7ef69f2694686179999d063ad452cfd5cf6cab.

So everything should be fine now.

Sebastian

Elchi3 commented:
Pulled from this branch again and will test this while editing MDN this week.

Elchi3 commented:
Let's merge this :tada: :tada: :tada: :tada: :tada:
